File: goofy-protocol-fis-be/src/main/java/com/masl/goofy_protocol_fis_be/crypto/symm/GlobSymmCrypto.java
package com.masl.goofy_protocol_fis_be.crypto.symm;

import com.masl.goofy_protocol_fis_be.crypto.asymm.*;

import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.util.Base64;
import java.util.List;
import java.util.regex.Pattern;

import static com.masl.goofy_protocol_fis_be.crypto.SecretUtils.ENC_DELIMITER;

public class GlobSymmCrypto {
    List<SymmCrypto> cryptoList = List.of(new SymmCryptoAES());

    public List<SymmCryptoType> getTypes() {
        return cryptoList.stream().map(SymmCrypto::getTypes).flatMap(List::stream).toList();
    }

    public SymmCrypto forType(SymmCryptoType type) {
        return cryptoList.stream().filter(c -> c.getTypes().contains(type)).findFirst().orElse(null);
    }

    public record ParsedEncData(byte[] data, SymmCryptoType type) {
        public static ParsedEncData parse(String value) {
            String[] parts = value.split(Pattern.quote(ENC_DELIMITER));
            if (parts.length != 2)
                throw new IllegalArgumentException("Invalid data format");
            SymmCryptoType type = SymmCryptoType.valueOf(parts[0]);
            byte[] data = Base64.getUrlDecoder().decode(parts[1]);
            return new ParsedEncData(data, type);
        }

        public String serialize() {
            return type.toString() + ENC_DELIMITER + Base64.getUrlEncoder().encodeToString(data);
        }
    }

    // Encrypt String into Parsed Data in Base64 Format
    public String encryptStr(String data, String secret, SymmCryptoType type) {
        return encrypt(data.getBytes(StandardCharsets.UTF_8), secret, type);
    }

    // Decrypt Parsed Data in Base64 Format into String
    public String decryptStr(String data, String secret) {
        return decrypt(data, secret) != null ? new String(decrypt(data, secret), StandardCharsets.UTF_8) : null;
    }

    // Encrypt Byte Array into Parsed Data in Base64 Format
    public String encrypt(byte[] data, String secret, SymmCryptoType type) {
        SymmCrypto crypto = forType(type);
        if (crypto == null || data == null || secret == null)
            throw new IllegalArgumentException("Invalid type, data or secret");
        byte[] res = crypto.encrypt(data, crypto.fromSecretString(secret, type), type);
        return new ParsedEncData(res, type).serialize();
    }

    // Decrypt Parsed Data in Base64 Format into Byte Array
    public byte[] decrypt(String data, String secret) {
        ParsedEncData parsed = ParsedEncData.parse(data);
        SymmCrypto crypto = forType(parsed.type());
        if (crypto == null || secret == null)
            throw new IllegalArgumentException("Invalid type or secret");
        return crypto.decrypt(parsed.data, crypto.fromSecretString(secret, parsed.type()), parsed.type());
    }

    // Encrypt Raw Byte Array into Byte Array (+ Crypto Type)
    public byte[] encryptRaw(byte[] data, String secret, SymmCryptoType type) {
        SymmCrypto crypto = forType(type);
        if (crypto == null || data == null || secret == null)
            throw new IllegalArgumentException("Invalid type, data or secret");
        byte[] res = crypto.encrypt(data, crypto.fromSecretString(secret, type), type);

        ByteBuffer buf = ByteBuffer.allocate(2 + res.length);
        buf.putShort(type.getValue());
        buf.put(res);
        return buf.array();
    }

    // Decrypt Raw Byte Array into Byte Array
    public byte[] decryptRaw(byte[] data, String secret) {
        if (data == null || secret == null || data.length < 2)
            throw new IllegalArgumentException("Invalid type or secret");

        ByteBuffer buf = ByteBuffer.wrap(data);
        SymmCryptoType type = SymmCryptoType.fromValue(buf.getShort());

        SymmCrypto crypto = forType(type);
        if (crypto == null)
            throw new IllegalArgumentException("Invalid type or secret");

        byte[] enc = new byte[buf.remaining()];
        buf.get(enc);

        return crypto.decrypt(enc, crypto.fromSecretString(secret, type), type);
    }


    // Default Methods using the Default Symmetric Crypto Algo
    public static final SymmCryptoType DEFAULT_TYPE = SymmCryptoType.AES_GCM_192;
    public byte[] encryptRaw(byte[] data, String secret) {return encryptRaw(data, secret, DEFAULT_TYPE);}
    public String encrypt(byte[] data, String secret) {return encrypt(data, secret, DEFAULT_TYPE);}
    public String encryptStr(String data, String secret) {return encryptStr(data, secret, DEFAULT_TYPE);}
}

File: goofy-protocol-fis-test-fe/libs/goofy-protocol-core/crypto/isolated/symm/glob-symm-crypto.mjs
import { ENC_DELIMITER } from "../secret-utils.mjs";
import { SymmCryptoType } from "./symm-crypto-type.mjs";

import {AES} from "./aes.mjs";

function bytesToB64Url(bytes) {
    let binary = "";
    const chunkSize = 0x8000;
    for (let i = 0; i < bytes.length; i += chunkSize) {
        binary += String.fromCharCode(...bytes.subarray(i, i + chunkSize));
    }
    return btoa(binary).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/g, "");
}

function b64UrlToBytes(str) {
    const s = str.replace(/-/g, "+").replace(/_/g, "/") + "===".slice((str.length + 3) % 4);
    const binary = atob(s);
    const out = new Uint8Array(binary.length);
    for (let i = 0; i < binary.length; i++) out[i] = binary.charCodeAt(i);
    return out;
}

function u16beToInt(x) {
    return ((x[0] & 0xff) << 8) | (x[1] & 0xff);
}
function intToU16be(n) {
    return new Uint8Array([(n >>> 8) & 0xff, n & 0xff]);
}

export class GlobSymmCrypto {
    constructor() {
        this.cryptoList = [
            new AES()
        ];
    }

    getTypes() {
        return this.cryptoList.flatMap(c => c.getTypes());
    }

    forType(type) {
        return this.cryptoList.find(c => c.getTypes().includes(type)) ?? null;
    }

    static ParsedEncData = class ParsedEncData {
        constructor(data, type) {
            this.data = data; // Uint8Array
            this.type = type; // SymmCryptoType
        }

        static parse(value) {
            const parts = value.split(ENC_DELIMITER);
            if (parts.length !== 2) throw new Error("Invalid data format");
            const type = SymmCryptoType[parts[0]];
            if (!type) throw new Error(`Unknown SymmCryptoType name: ${parts[0]}`);
            const data = b64UrlToBytes(parts[1]);
            return new GlobSymmCrypto.ParsedEncData(data, type);
        }

        serialize() {
            return `${this.type.name}${ENC_DELIMITER}${bytesToB64Url(this.data)}`;
        }
    };

    async encryptStr(data, secret, type = GlobSymmCrypto.DEFAULT_TYPE) {
        return this.encrypt(new TextEncoder().encode(data), secret, type);
    }

    async decryptStr(data, secret) {
        const bytes = await this.decrypt(data, secret);
        return bytes !== null ? new TextDecoder().decode(bytes) : null;
    }

    async encrypt(dataBytes, secret, type = GlobSymmCrypto.DEFAULT_TYPE) {
        const crypto = this.forType(type);
        if (!crypto || dataBytes == null || secret == null) throw new Error("Invalid type, data or secret");
        const res = await crypto.encrypt(dataBytes, await crypto.fromSecretString(secret, type), type);
        return new GlobSymmCrypto.ParsedEncData(res, type).serialize();
    }

    async decrypt(data, secret) {
        const parsed = GlobSymmCrypto.ParsedEncData.parse(data);
        const crypto = this.forType(parsed.type);
        if (!crypto || secret == null) throw new Error("Invalid type or secret");
        return crypto.decrypt(parsed.data, await crypto.fromSecretString(secret, parsed.type), parsed.type);
    }

    async encryptRaw(dataBytes, secret, type = GlobSymmCrypto.DEFAULT_TYPE) {
        const crypto = this.forType(type);
        if (!crypto || dataBytes == null || secret == null) throw new Error("Invalid type, data or secret");

        const res = await crypto.encrypt(dataBytes, await crypto.fromSecretString(secret, type), type);

        // Java: ByteBuffer.allocate(2 + res.length); buf.putShort(type.getValue()); buf.put(res)
        // -> big-endian u16
        const out = new Uint8Array(2 + res.length);
        const u16 = intToU16be(type.value);
        out.set(u16, 0);
        out.set(res, 2);
        return out;
    }

    async decryptRaw(dataBytes, secret) {
        if (!dataBytes || secret == null || dataBytes.length < 2) throw new Error("Invalid type or secret");

        const typeValue = u16beToInt(dataBytes.subarray(0, 2));
        const type = SymmCryptoType.fromValue(typeValue); // you provide this method in the enum module

        const crypto = this.forType(type);
        if (!crypto) throw new Error("Invalid type or secret");

        const enc = dataBytes.subarray(2);
        return crypto.decrypt(enc, await crypto.fromSecretString(secret, type), type);
    }

    static DEFAULT_TYPE = SymmCryptoType.AES_GCM_256;
}